azure-pipelines.yml

```diff
@@ -202,6 +202,14 @@ jobs:
     parameters:
       trigger_sha: '$(trigger_sha)'
 
+- job: compat_stable_protobuf_pr
```

Comment:
I’m confused. Do you not want to run this on PRs? With this, we’ll end up not noticing it on PRs and then we get a red main.

Reply:
The current change only ensures that we can't release if we break protobuf, which is a product-level requirement.

The per-PR-commit check, used during development, will be a subsequent change and will check against the target branch's HEAD (and it will replace the current manual check in `fmt.sh`).

Reply:
It doesn’t just block releases though, it results in a red main and breaks our CI completely. I’d rather add the per PR checks first, otherwise this will result in a lot of noise and broken CI.

Reply:
> breaks our CI completely

What does it mean? Consider that we allow making exceptions in per-PR checks, so it will still be possible to merge them.

Reply:
Having to restore compatibility in the rare case where we had a merge conflict is fine. Having to restore compatibility with a check that we didn’t even run on a PR is imho asking for this to be too frequent of an occurence.

Reply:
Agree with @cocreature that it'd be better if we merged the per-commit one first, though I'm not _too_ concerned about merging this one first. Yes, there's a risk we'd only catch issues on main, and that would be annoying as then we have red commits on main, but:
1. I'm assuming (hoping?) we don't regularly randomly break protobufs and only rely on CI to tell us about it, so the chance of actually breaking main are not that high.
2. I'm also assuming the PR check is the very next item on your todo list so the window of opportunity is not that large.
3. As this is a check that is only run on main commits, and thus not on PRs, this being red will not block merges.
4. As this is not added to `collect_build_data` dependencies, it will not result in any Slack message.

🤔 

Actually, that last one is kinda bad. What _should_ happen if that test fails? Currently, it will just silently put a red cross in the commit view on GitHub, which is probably not all that useful. Should this perhaps send a Slack message to the team responsible for protobufs?

Reply:
> Suppose we use the override in a PR though, because we decide to temporarily allow incompatibility; is it problematic to have the CI failing on `main` on every PR merge until we restore compatibility?

My understanding was that we wanted the option for individual commits to break compatibility _with previous commits_ in-between releases, not the compatibility with the latest release. So I would still expect this to be green on main commits, because on main we only check with latest release. Did I misunderstand something? (Also that does mean this should probably also run on PRs and thus be added to `ci/build.yml` instead of `azure-pipelines.yml`.)

Reply:
You are right about 1 and 2 and thanks for confirming 3.

As for 4, there is no single team responsible for all protobufs, some are shared (see `CODEOWNERS` for more details) and anyway work across ownership has happened lately and will likely happen again. Given that, what would you consider a useful Slack message? E.g., is sending one to the whole team better than silence?

Reply:
yeah tell-slack-failed was what I was thinking off

Reply:
> My understanding was that we wanted the option for individual commits to break compatibility with previous commits in-between releases

Yes, you are right. I wonder if there is a case where we would like to temporarily break compatibility with the latest stable but I can't think of any. I'll move the check to the PR and we can always decide to be less strict later on.

Reply:
I think having the check (also) in the PR means that it should take into account not only `main` as a target but also releases, is that correct? This means that the most recent stable tag is now relative again.

Reply:
If I understand correctly, this check will be run on:

1. `main` commits and PRs targeting `main`: we want `git tag --no-merged` because release tags happen on release branches (that are not reachable from `main`). Or simply `git tag` because we're considering the most recent anyway.
2. `release/*` commits and PRs targeting `release/*`: we want `git tag --merged` because the most recent release tag for that release branch is reachable from it.

So we need a way to distinguish between these two cases. Figuring out the merge commits is not a problem but what about the PRs'? Do we have the target branch available?

Reply:
Assuming my reasoning is correct, I think we can probably leverage [`System.PullRequest.TargetBranch`](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables).

Reply:
> So I'd say --merged for PRs to release branches, and nothing for main PRs.
👍🏼 
> [...] we mostly only backport from main into release branches

It's a good idea to make this assumption explicit, I'll add a comment about that.

Reply:
all good, thanks!